File: test/examples/cryptarithm_sat/cryptarithm_sat.cc




























#include "ortools/sat/cp_model.h"

namespace operations_research {
namespace sat {

void SendMoreMoney() {
  CpModelBuilder cp_model;

  

  Domain all_digits(0, 9);
  Domain non_zero_digits(1, 9);
  

  

  const IntVar s = cp_model.NewIntVar(non_zero_digits);
  const IntVar e = cp_model.NewIntVar(all_digits);
  const IntVar n = cp_model.NewIntVar(all_digits);
  const IntVar d = cp_model.NewIntVar(all_digits);
  

  const IntVar m = cp_model.NewIntVar(non_zero_digits);
  const IntVar o = cp_model.NewIntVar(all_digits);
  const IntVar r = cp_model.NewIntVar(all_digits);
  const IntVar y = cp_model.NewIntVar(all_digits);

  

  

  const BoolVar c0 = cp_model.NewBoolVar();
  const BoolVar c1 = cp_model.NewBoolVar();
  const BoolVar c2 = cp_model.NewBoolVar();
  const BoolVar c3 = cp_model.NewBoolVar();

  

  cp_model.AddAllDifferent({s, e, n, d, m, o, r, y});

  

  cp_model.AddEquality(c0, m);

  

  cp_model.AddEquality(c1 + s + m, o + 10 * c0);

  

  cp_model.AddEquality(c2 + e + o, n + 10 * c1);

  

  cp_model.AddEquality(c3 + n + r, e + 10 * c2);

  

  cp_model.AddEquality(d + e, y + 10 * c3);

  

  const CpSolverResponse response = Solve(cp_model.Build());
  LOG(INFO) << CpSolverResponseStats(response);
  LOG(INFO) << "s: " << SolutionIntegerValue(response, s);
  LOG(INFO) << "e: " << SolutionIntegerValue(response, e);
  LOG(INFO) << "n: " << SolutionIntegerValue(response, n);
  LOG(INFO) << "d: " << SolutionIntegerValue(response, d);
  LOG(INFO) << "m: " << SolutionIntegerValue(response, m);
  LOG(INFO) << "o: " << SolutionIntegerValue(response, o);
  LOG(INFO) << "r: " << SolutionIntegerValue(response, r);
  LOG(INFO) << "y: " << SolutionIntegerValue(response, y);
}

}  

}  


int main() {
  operations_research::sat::SendMoreMoney();
  return EXIT_SUCCESS;
}

File: test/ok/cryptarithm_sat.ts
import { operations_research as op } from '../../src'


function test()
{
    let cp_model = new op.sat.CpModelBuilder();
    let all_digits = new op.Domain(0, 9);
    let non_zero_digits = new op.Domain(1, 9);

    let s = cp_model.NewIntVar(non_zero_digits);
    let e = cp_model.NewIntVar(all_digits);
    let n = cp_model.NewIntVar(all_digits);
    let d = cp_model.NewIntVar(all_digits);

    let m = cp_model.NewIntVar(non_zero_digits);
    let o = cp_model.NewIntVar(all_digits);
    let r = cp_model.NewIntVar(all_digits);
    let y = cp_model.NewIntVar(all_digits);

    let c0 = cp_model.NewBoolVar();
    let c1 = cp_model.NewBoolVar();
    let c2 = cp_model.NewBoolVar();
    let c3 = cp_model.NewBoolVar();

    cp_model.AddAllDifferent([s, e, n, d, m, o, r, y]);

    cp_model.AddEquality(c0, m);

    let lexpr = op.sat.operator_plus(c1, s);
    lexpr = op.sat.operator_plus(lexpr, m);
    let rexpr = op.sat.operator_plus(o, op.sat.operator_times(10, c0))
    cp_model.AddEquality(lexpr, rexpr);

    lexpr = op.sat.operator_plus(c2, e);
    lexpr = op.sat.operator_plus(lexpr, o);
    rexpr = op.sat.operator_plus(n, op.sat.operator_times(10, c1))
    cp_model.AddEquality(lexpr, rexpr);

    lexpr = op.sat.operator_plus(c3, n);
    lexpr = op.sat.operator_plus(lexpr, r);
    rexpr = op.sat.operator_plus(e, op.sat.operator_times(10, c2))
    cp_model.AddEquality(lexpr, rexpr);

    lexpr = op.sat.operator_plus(d, e);
    rexpr = op.sat.operator_plus(y, op.sat.operator_times(10, c3))
    cp_model.AddEquality(lexpr, rexpr);

    let response = op.sat.Solve(cp_model.Build());
    console.log(op.sat.CpSolverResponseStats(response));
    console.log("s: " + op.sat.SolutionIntegerValue(response, s));
    console.log("e: " + op.sat.SolutionIntegerValue(response, e));
    console.log("n: " + op.sat.SolutionIntegerValue(response, n));
    console.log("d: " + op.sat.SolutionIntegerValue(response, d));
    console.log("m: " + op.sat.SolutionIntegerValue(response, m));
    console.log("o: " + op.sat.SolutionIntegerValue(response, o));
    console.log("r: " + op.sat.SolutionIntegerValue(response, r));
    console.log("y: " + op.sat.SolutionIntegerValue(response, y));
}
test();
